### PR TITLE
feat(gizmo): custom colors for gizmos

### PR DESCRIPTION
**Pull request by PhilippeMorier**
### What

- support setting custom colors for hover & disable material for
  - plane-drag gizmo
  - plane-rotation gizmo
- related https://forum.babylonjs.com/t/setts-of-gizmo-color-when-events/22271

### Why

- to avoid setting it via accessing "private" members
- to avoid white reflection color on manually set hover material  
  image

### Visuals

|Before [PG](https://playground.babylonjs.com/#VJ0MJD)|After|
|-|-|
|babylonjs-gizmo-before|babylonjs-gizmo-after|

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14525/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14525/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14525/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14525/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14525/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14525/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14525/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14525/merge#BCU1XR#0

File: packages/dev/core/src/Gizmos/planeRotationGizmo.ts [hunk 1]
```diff
@@ -122,34 +122,37 @@ export class PlaneRotationGizmo extends Gizmo implements IPlaneRotationGizmo {
...
+                start += base;
+                end += base;
+
+                float intensity = 0.;
+                for (int i = 0; i < 5; i++)
+                {
+                    intensity += max(step(start, angle) - step(end, angle), 0.);
+                    angle += twopi;
+                }
+                gl_FragColor = vec4(${color.asArray()}, min(intensity * 0.25, 0.8)) * opacity;
```
Comment (RaananW) [inline packages/dev/core/src/Gizmos/planeRotationGizmo.ts:152]: as array generates a new array on every call. To avoid garbage collection it would be better to reference the components directly:
```suggestion
                gl_FragColor = vec4(${color.r}, ${color.g}, ${color.b}, min(intensity * 0.25, 0.8)) * opacity;
```

**Review by RaananW — COMMENTED** [review]
I like the change! I will let Cedric go over it as well.

I just have a single comment about the asArray call. otherwise - make sure to run `npm run format:fix`  to make sure your code is formatted

File: packages/dev/core/src/Gizmos/planeRotationGizmo.ts [hunk 2]
```diff
@@ -122,34 +122,37 @@ export class PlaneRotationGizmo extends Gizmo implements IPlaneRotationGizmo {
...
-            float base = abs(floor(start / twopi)) * twopi;
-            start += base;
-            end += base;
-
-            float intensity = 0.;
-            for (int i = 0; i < 5; i++)
-            {
-                intensity += max(step(start, angle) - step(end, angle), 0.);
-                angle += twopi;
+    protected static _createRotationGizmoFragmentShader(color: Color3) {
```
Comment (Popov72) [inline packages/dev/core/src/Gizmos/planeRotationGizmo.ts:125]: Technically it is a breaking change if someone subclassed this class and replaced `_RotationGizmoFragmentShader`...

To be safe, you could do something like:
```typescript
protected static _RotationGizmoFragmentShader: Nullable<string> = null;
...
Effect.ShadersStore["rotationGizmoFragmentShader"] = PlaneRotationGizmo._RotationGizmoFragmentShader ?? PlaneRotationGizmo._createRotationGizmoFragmentShader(hoverColor);
```
But I'm not sure how many people would have subclassed `PlaneRotationGizmo` in the first place...

File: packages/dev/core/src/Gizmos/planeRotationGizmo.ts (hunk 1 above)
Comment (Popov72) [reply packages/dev/core/src/Gizmos/planeRotationGizmo.ts:152]: Also, you should make sure that the stringification of r/g/b always has a decimal point, else it will generate an error at shader compilation time. For eg, "1,1,0" won't work, it should be "1.,1.,0." (or "1.0,1.0,0.0").

**Review by Popov72 — CHANGES_REQUESTED** [review]

File: packages/dev/core/src/Gizmos/planeRotationGizmo.ts [hunk 3]
```diff
@@ -122,34 +122,45 @@ export class PlaneRotationGizmo extends Gizmo implements IPlaneRotationGizmo {
...
+        let r = `${color.r}`;
+        let g = `${color.g}`;
+        let b = `${color.b}`;
+
+        r = r.includes('.') ? r : `${r}.`;
+        g = g.includes('.') ? g : `${g}.`;
+        b = b.includes('.') ? b : `${b}.`;
```
Comment (PhilippeMorier) [inline packages/dev/core/src/Gizmos/planeRotationGizmo.ts:132]: This feels a bit a lot of code to achieve having always a "." in the color :innocent:.

Does maybe someone have a better idea?

File: packages/dev/core/src/Gizmos/planeRotationGizmo.ts [hunk 4]
```diff
@@ -181,24 +202,32 @@ export class PlaneRotationGizmo extends Gizmo implements IPlaneRotationGizmo {
...
+            },
+            this.gizmoLayer.utilityLayerScene
+        );
         this._rotationDisplayPlane.rotation.z = Math.PI * 0.5;
         this._rotationDisplayPlane.parent = this._gizmoMesh;
         this._rotationDisplayPlane.setEnabled(false);
 
         Effect.ShadersStore["rotationGizmoVertexShader"] = PlaneRotationGizmo._RotationGizmoVertexShader;
-        Effect.ShadersStore["rotationGizmoFragmentShader"] = PlaneRotationGizmo._RotationGizmoFragmentShader;
+        Effect.ShadersStore["rotationGizmoFragmentShader"] = PlaneRotationGizmo._createRotationGizmoFragmentShader(hoverColor);
```
Comment (RaananW) [inline packages/dev/core/src/Gizmos/planeRotationGizmo.ts:230]: I was thinking about this a bit. Setting the shader as a fixed string will pevent it from being reused with different colors.
setting the color as a uniform to the shader material will allow multiple instances of this to use different colors

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14525/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/core/src/Gizmos/planeRotationGizmo.ts (hunk 4 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Gizmos/planeRotationGizmo.ts:230]: I second Raanan. Standard way is to use uniform, not to change the shader code.

Comment (RaananW) [reply packages/dev/core/src/Gizmos/planeRotationGizmo.ts:230]: the solution with adding a floating point will also be solved if you do that.

Comment (CedricGuillemet) [reply packages/dev/core/src/Gizmos/planeRotationGizmo.ts:230]: Sure! Check this line in the shader: `uniform vec3 angles;`  It's a uniform constant that declared in the shader and set on the CPU side in 2 steps:
- declaring it here : https://github.com/BabylonJS/Babylon.js/blob/f7793c25a456b56801e6b352287341b0a1d4525c/packages/dev/core/src/Gizmos/planeRotationGizmo.ts#L240
- and set here : https://github.com/BabylonJS/Babylon.js/blob/f7793c25a456b56801e6b352287341b0a1d4525c/packages/dev/core/src/Gizmos/planeRotationGizmo.ts#L278

Basically, you have to do the same thing with diffuse colors instead of angles.

Comment (PhilippeMorier) [reply packages/dev/core/src/Gizmos/planeRotationGizmo.ts]: Thanks for the hint!

I think/hope I got it :tada:.

**Review by CedricGuillemet — APPROVED** [review]
LGTM

**Review by RaananW — APPROVED** [review]

**Review by Popov72 — APPROVED** [review]
LGTM!

If you want, you could add getters and setters for the new colors, so that they can be changed at any time.

**Comment by RaananW** [comment]
i'll merge this. @PhilippeMorier - thank you so much for your time and contribution! If you want to add the ability to change the colors in runtime (which is now possible because of the shader changes) you can submit another PR.
